File: servidor.java
import java.io.*;
import java.net.*;

public class Servidor {
    public static void main(String[] args) {
        ServerSocket servidor = null;
        Socket cliente = null;
        BufferedReader entrada = null;
        PrintWriter salida = null;
        
        try {
            // Crear un socket servidor en el puerto 5000
            servidor = new ServerSocket(5000);
            System.out.println("Servidor iniciado, esperando cliente...");
            
            // Aceptar conexiones del cliente
            cliente = servidor.accept();
            System.out.println("Cliente conectado");

            // Flujos para recibir y enviar mensajes
            entrada = new BufferedReader(new InputStreamReader(cliente.getInputStream()));
            salida = new PrintWriter(cliente.getOutputStream(), true);
            
            // Ciclo para intercambiar mensajes con el cliente
            String mensaje;
            while ((mensaje = entrada.readLine()) != null) {
                System.out.println("Cliente: " + mensaje);
                salida.println("Mensaje recibido: " + mensaje);
            }
        } catch (IOException e) {
            System.out.println("Error en el servidor: " + e.getMessage());
        } finally {
            try {
                if (entrada != null) entrada.close();
                if (salida != null) salida.close();
                if (cliente != null) cliente.close();
                if (servidor != null) servidor.close();
            } catch (IOException e) {
                System.out.println("Error cerrando recursos: " + e.getMessage());
            }
        }
    }
}

File: servidor.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

int main() {
    int servidor, cliente, len, numero;
    struct sockaddr_in servidor_addr, cliente_addr;

    // Crear socket
    servidor = socket(AF_INET, SOCK_STREAM, 0);
    if (servidor == -1) {
        printf("Error al crear el socket\n");
        return -1;
    }

    // Configurar la estructura del servidor
    servidor_addr.sin_family = AF_INET;
    servidor_addr.sin_addr.s_addr = INADDR_ANY;
    servidor_addr.sin_port = htons(5000);

    // Asociar socket con la dirección IP y el puerto
    if (bind(servidor, (struct sockaddr *)&servidor_addr, sizeof(servidor_addr)) < 0) {
        printf("Error al asociar el socket\n");
        close(servidor);
        return -1;
    }

    // Escuchar conexiones entrantes
    listen(servidor, 3);
    printf("Esperando conexión...\n");

    len = sizeof(cliente_addr);
    cliente = accept(servidor, (struct sockaddr *)&cliente_addr, (socklen_t*)&len);
    if (cliente < 0) {
        printf("Error al aceptar la conexión\n");
        close(servidor);
        return -1;
    }

    printf("Cliente conectado\n");

    // Intercambio de números enteros
    while (1) {
        // Recibir número del cliente
        if (recv(cliente, &numero, sizeof(int), 0) > 0) {
            printf("Número recibido del cliente: %d\n", numero);
            // Enviar el mismo número de vuelta
            if (send(cliente, &numero, sizeof(int), 0) < 0) {
                printf("Error al enviar el número\n");
                break;
            }
        } else {
            printf("Cliente desconectado\n");
            break;
        }
    }

    close(cliente);
    close(servidor);
    return 0;
}
